ports-of-call/portable_arrays.hpp

```diff
@@ -24,438 +24,275 @@
-  PORTABLE_FORCEINLINE_FUNCTION int GetDim4() const { return nx4_; }
-  PORTABLE_FORCEINLINE_FUNCTION int GetDim5() const { return nx5_; }
-  PORTABLE_FORCEINLINE_FUNCTION int GetDim6() const { return nx6_; }
+  template <std::size_t I>
+  PORTABLE_FUNCTION constexpr auto GetDim() const {
+    return nxs_[rank_ - I];
+  }
+
+  // legacy API, TODO: deprecate
+  [[deprecated("Use GetDim<N> instead.")]] PORTABLE_FORCEINLINE_FUNCTION int
+  GetDim1() const {
+    return GetDim<1>();
+  }
+  [[deprecated("Use GetDim<N> instead.")]] PORTABLE_FORCEINLINE_FUNCTION int
+  GetDim2() const {
+    return GetDim<2>();
+  }
+  [[deprecated("Use GetDim<N> instead.")]] PORTABLE_FORCEINLINE_FUNCTION int
+  GetDim3() const {
+    return GetDim<3>();
+  }
+  [[deprecated("Use GetDim<N> instead.")]] PORTABLE_FORCEINLINE_FUNCTION int
+  GetDim4() const {
+    return GetDim<4>();
+  }
+  [[deprecated("Use GetDim<N> instead.")]] PORTABLE_FORCEINLINE_FUNCTION int
+  GetDim5() const {
+    return GetDim<5>();
+  }
+  [[deprecated("Use GetDim<N> instead.")]] PORTABLE_FORCEINLINE_FUNCTION int
+  GetDim6() const {
+    return GetDim<6>();
+  }
   PORTABLE_INLINE_FUNCTION int GetDim(size_t i) const {
     // TODO: remove if performance cirtical
     assert(0 < i && i <= 6 && "PortableMDArrays are max 6D");
     switch (i) {
     case 1:
-      return GetDim1();
+      return GetDim<1>();
     case 2:
-      return GetDim2();
+      return GetDim<2>();
     case 3:
-      return GetDim3();
+      return GetDim<3>();
     case 4:
-      return GetDim4();
+      return GetDim<4>();
     case 5:
-      return GetDim5();
+      return GetDim<5>();
     case 6:
-      return GetDim6();
+      return GetDim<6>();
     }
     return -1;
   }
 
-  // a function to get the total size of the array
-  PORTABLE_FORCEINLINE_FUNCTION int GetSize() const {
-    return nx1_ * nx2_ * nx3_ * nx4_ * nx5_ * nx6_;
+  PORTABLE_FORCEINLINE_FUNCTION int GetSize() const noexcept {
+    return util::array_reduce(nxs_, 1, std::multiplies<size_type>{});
   }
-  PORTABLE_FORCEINLINE_FUNCTION std::size_t GetSizeInBytes() const {
-    return nx1_ * nx2_ * nx3_ * nx4_ * nx5_ * nx6_ * sizeof(T);
-  }
-
-  PORTABLE_INLINE_FUNCTION size_t GetRank() const {
-    for (int i = 6; i >= 1; i--) {
-      if (GetDim(i) > 1) return i;
-    }
-    return 0;
+  PORTABLE_FORCEINLINE_FUNCTION size_type GetSizeInBytes() const noexcept {
+    return GetSize() * sizeof(T);
   }
 
-  PORTABLE_INLINE_FUNCTION void Reshape(int nx6, int nx5, int nx4, int nx3, int nx2,
-                                        int nx1) {
-    assert(nx6 * nx5 * nx4 * nx3 * nx2 * nx1 == GetSize());
-    nx1_ = nx1;
-    nx2_ = nx2;
-    nx3_ = nx3;
-    nx4_ = nx4;
-    nx5_ = nx5;
-    nx6_ = nx6;
+  PORTABLE_INLINE_FUNCTION size_type GetRank() const noexcept { return rank_; }
+  template <typename... NXs>
+  PORTABLE_INLINE_FUNCTION void Reshape(NXs... nxs) {
+    assert(util::array_reduce(IArray<D>{nxs...}, 1, std::multiplies<size_type>{}) ==
```

Comment:
I know this was an assert previously, but now we have `PORTABLE_REQUIRE`. Let's use that instead as it will provide more useful/verbose output.